### DISCUSSION TITLE
In the future maybe reconsider introducing warnings message for future breaking changes

### DISCUSSION THREAD

**Discussion by DaveWelling** [statelyai/xstate#3621] [General]
I want to acknowledge first that communicating future breaking changes is an important and very considerate action.  It demonstrates contributors are really thinking about how future users will engage with xstate.
Recently somewhere in 4.33 a warning was added to xstate that a future version would be breaking and that users should opt into the potentially breaking behavior by using the predictableActionArguments flag.  This is actually very disruptive since it floods unit testing output with this warning, obscuring more critical information in that output.
I'm sure you're aware many devs are under tight deadlines and refactoring the codebase to accommodate this change is something that needs to be added to the backlog and triaged.
For this reason, I was forced to set XState at v4.32.1 to avoid the warning.
Since semver implies breaking changes when a major release occurs, increasing the major version is probably enough under most circumstances to communicate this type of change, but if additional communication seems necessary, there are also informational messages which can be added to module install actions like `npm install`.

**Comment by Andarist** [comment]
Hi there, thank you for the comment and the reasonable critique. I'd like to address a couple of points raised here.

> Recently somewhere in 4.33 a warning was added to xstate that a future version would be breaking and that users should opt into the potentially breaking behavior by using the predictableActionArguments flag. This is actually very disruptive since it floods unit testing output with this warning, obscuring more critical information in that output.

I've intentionally made this to only log once and not for each machine missing this flag. However, I've missed the fact that with the testing framework the modules are often reinstantiated per test file and thus the log can be still logged multiple times during a single "run".

I've seen worse log floods in the past ([React deprecating some lifecycle methods](https://pl.reactjs.org/blog/2018/03/29/react-v-16-3.html#component-lifecycle-changes)) but perhaps that was a lesson for us what should be avoided.

> I'm sure you're aware many devs are under tight deadlines and refactoring the codebase to accommodate this change is something that needs to be added to the backlog and triaged.

Yes. That's fine though - everybody should update at their own pace.

> For this reason, I was forced to set XState at v4.32.1 to avoid the warning.

This is presented as a bad thing but I don't see why it's viewed as such. This is the exact reason why versions exist and why we can pin them. We are not forcing anybody to update ASAP, everyone should do that at their own pace.

> there are also informational messages which can be added to module install actions like npm install.

Let's be honest - a lot of people don't read those at all. I stand by the decision to make this a runtime log over the installation log. I agree that maybe more things could have been considered and that we could have introduced a way to disable this warning globally or something.

**Comment by DaveWelling** [comment]
It's amazing to me that you responded so quickly, thoroughly and thoughtfully.  Thank you!  I think that adding a flag to disable similar warnings would completely fix the testing log problems.
I could go into greater detail about how semver is designed to handle this situation, but I think you are probably complete cognizant of those things.  I will say that pinning to a minor version precludes the ability to receive non-breaking fixes (such as any critical vulnerability fixes) during the natural build cycle. 
Perhaps the most compelling thing I could ask is: if this is the appropriate way to announce future breaking changes, why doesn't everybody do this, and what would it look like if they did?
If you still remain firm in your opinion, my feeling is that your opinion should far, far outweigh my own because you are the one investing your time and care into the software!  :)  
Thanks again!
